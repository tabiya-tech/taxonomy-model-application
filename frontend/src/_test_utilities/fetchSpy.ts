export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
