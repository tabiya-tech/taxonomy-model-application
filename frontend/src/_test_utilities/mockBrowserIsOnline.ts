import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
