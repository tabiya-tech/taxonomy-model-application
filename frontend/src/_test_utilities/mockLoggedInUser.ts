import { AuthContextValue, TabiyaUser, TabiyaUserRole, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>){
  let user: TabiyaUser = mockedValues.user || { username: "foo", roles: [] };
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: TabiyaUserRole) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues
  });
}
