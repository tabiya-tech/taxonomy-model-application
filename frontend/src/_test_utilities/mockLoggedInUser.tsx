import {
  AuthContextValue,
  TabiyaUser,
  authContextDefaultValue,
  AuthContext,
  AuthProviderProps,
} from "src/auth/AuthProvider";
import AuthAPISpecs from "api-specifications/auth";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>) {
  let user: TabiyaUser = mockedValues.user || TestUsers.Anonymous;

  const value = {
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: AuthAPISpecs.Enums.TabiyaRoles) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues,
  };

  // @ts-ignore
  jest.spyOn(require("src/auth/AuthProvider"), "AuthProvider").mockImplementation((props: AuthProviderProps) => {
    return <AuthContext.Provider value={value}>{props.children}</AuthContext.Provider>;
  });
}

export const TestUsers = {
  ModelManager: {
    username: "ModelManagerUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER],
  },
  RegisteredUser: {
    username: "RegisteredUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER],
  },
  Anonymous: {
    username: "AnonymousUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS],
  },
};
