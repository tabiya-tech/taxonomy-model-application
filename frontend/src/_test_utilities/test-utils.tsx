// Based on https://testing-library.com/docs/react-testing-library/setup/

import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/providers";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
