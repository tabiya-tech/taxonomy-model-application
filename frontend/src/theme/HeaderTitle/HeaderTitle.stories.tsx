import HeaderTitle from "./HeaderTitle";
import { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof HeaderTitle> = {
  title: "Components/HeaderTitle",
  component: HeaderTitle,
  tags: ["autodocs"],
  argTypes: {},
};

export default meta;

type Story = StoryObj<typeof HeaderTitle>;

export const Shown: Story = {
  args: {
    children: "Header Title",
  },
};
