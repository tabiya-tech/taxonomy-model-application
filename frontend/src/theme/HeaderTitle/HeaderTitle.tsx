import React from "react";
import { Typography } from "@mui/material";

export interface HeaderTitleProps {
  children: React.ReactNode;
  "data-testid"?: string;
}

/**
 * Reusable title component for Headers in the app.
 * Should use h2 variant of the mui typography
 * @param props gets data-testid for testing from parent
 * @constructor
 */
const HeaderTitle: React.FC<HeaderTitleProps> = (props: Readonly<HeaderTitleProps>) => {
  return (
    <Typography variant="h2" data-testid={props["data-testid"]} sx={{ wordWrap: "break-word" }}>
      {props.children}
    </Typography>
  );
};

export default HeaderTitle;
