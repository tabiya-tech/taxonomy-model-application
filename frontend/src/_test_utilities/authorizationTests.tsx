import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
