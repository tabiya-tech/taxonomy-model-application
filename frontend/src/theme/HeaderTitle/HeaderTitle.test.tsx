// mute the console
import "src/_test_utilities/consoleMock";

import HeaderTitle from "./HeaderTitle";
import { render, screen } from "src/_test_utilities/test-utils";

describe("HeaderTitle", () => {
  beforeEach(() => {
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
  });

  test("should render header component with provided title", () => {
    // GIVEN the title
    const givenTitle = "Header title";

    // WHEN HeaderTitle component is rendered
    render(<HeaderTitle data-testid="header-title">{givenTitle}</HeaderTitle>);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the header title is in the document with the data-testid
    const actualElement = screen.getByTestId("header-title");
    expect(actualElement).toBeInTheDocument();
    // AND the header title has the given title
    expect(actualElement).toHaveTextContent(givenTitle);
    // AND to match the snapshot
    expect(actualElement).toMatchSnapshot();
  });
});
