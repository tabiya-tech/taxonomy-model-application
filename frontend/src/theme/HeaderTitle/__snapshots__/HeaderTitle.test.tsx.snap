// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderTitle should render header component with provided title 1`] = `
<h2
  class="MuiTypography-root MuiTypography-h2 css-ca93eg-MuiTypography-root"
  data-testid="header-title"
>
  Header title
</h2>
`;
